HPCC-15047 Refactor HTPass into an embedded SecManager plugin

RussWhitehead opened the pull request:
Refactor HTPasswrd security manager to be a true plugin. Add files for
configmanager to gather common plugin artifacts and custom HTPassword
specific

Signed-off-by: Russ Whitehead william.whitehead@lexisnexis.com

Signed-off-by: Russ Whitehead william.whitehead@lexisnexis.com

RussWhitehead committed:
HPCC-15047 Refactor HTPass into an embedded SecManager plugin

Refactor HTPasswrd security manager to be a true plugin. Add files for
configmanager to gather common plugin artifacts and custom HTPassword
specific

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

RussWhitehead commented:
@rpastrana Please review 
@garonsky Please review the configuration changes

garonsky commented on system/security/plugins/htpasswdSecurity/initfiles/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+#    limitations under the License.
+################################################################################
+
+cmake_minimum_required(VERSION 2.8)
```
I think we require 3.x

RussWhitehead replied:
Searching the HPCC-Platform tree I found this. Maybe under a different Jira we should fix them all?
  C:\HPCC\HPCC-Platform\CMakeLists.txt(50):cmake_minimum_required(VERSION 2.8.11)
  C:\HPCC\HPCC-Platform\docs\CMakeLists.txt(17):cmake_minimum_required(VERSION 2.8)
  C:\HPCC\HPCC-Platform\initfiles\CMakeLists.txt(17):cmake_minimum_required(VERSION 2.8)
  C:\HPCC\HPCC-Platform\lib2\CMakeLists.txt(17):cmake_minimum_required(VERSION 2.8)

richardkchapman replied:
Good idea - did you open the Jira?

RussWhitehead replied:
https://track.hpccsystems.com/browse/HPCC-15213

rpastrana commented on esp/bindings/http/platform/httpbinding.cpp:
```diff
@@ -186,7 +186,7 @@ EspHttpBinding::EspHttpBinding(IPropertyTree* tree, const char *bindname, const
                 {
                     Owned<IPropertyTree> secMgrs;
-                    VStringBuffer sm("SecurityManagers/SecurityManager[@name='%s']", m_authmethod.str());
+                    VStringBuffer sm("SecurityManagers/SecurityManager[@name='%s']/HtpasswdSecurityManager", m_authmethod.str());
```
We shouldn't hardcode plugin names in the platform. We should provide the SecurityManager tree rooted at the SecurityManager element, or we might be kind to the plug-in developers and provide the tree rooted at the first child of the SecurityManager tree where @name = m_authmethod.str()

rpastrana commented on system/security/plugins/htpasswdSecurity/htpasswdSecurity.cpp:
```diff
@@ -25,11 +25,13 @@
-		if (authconfig)
-			authconfig->getProp("@htpasswdFile", pwFile);
-		apr_initialized = false;
+        if (secMgrCfg)
```
it's hard to tell in the github diff, but indent might be off at these lines

RussWhitehead replied:
Its a tab thing, the legacy was tabbed and the new stuff wasn't so it looked horrible. I didn't want to untabify the whole thing so I just did the body of the constructor

rpastrana commented on system/security/plugins/htpasswdSecurity/htpasswdSecurity.cpp:
```diff
@@ -25,11 +25,13 @@
 {
 public:
-    CHtpasswdSecurityManager(const char *serviceName, IPropertyTree *authconfig) : CBaseSecurityManager(serviceName, (IPropertyTree *)NULL)
+    CHtpasswdSecurityManager(const char *serviceName, IPropertyTree *secMgrCfg, IPropertyTree *authconfig) : CBaseSecurityManager(serviceName, (IPropertyTree *)NULL)
```
nitpick but some arguments use camel, one all lowers...

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on system/security/plugins/htpasswdSecurity/initfiles/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+################################################################################
```
If these are new files, the year might need to be updated.

rpastrana commented on system/security/plugins/htpasswdSecurity/initfiles/componentfiles/configxml/buildset.xml:
```diff
@@ -0,0 +1,32 @@
+      <BuildSet deployable="no"
+                installSet="deploy_map.xml"
+                name="htpasswdsecmgr"
+                path="componentfiles/htpasswdsecmgr"
```
If the buildset "htpasswdsecmgr" existed previously, I would consider changing the name of the newly created plugin to avoid any name collisions (even though it is unlikely to happen)

rpastrana commented:
@RussWhitehead reviewed

richardkchapman commented:
Needs rebasing for a clean merge

RussWhitehead committed:
Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

RussWhitehead commented:
@rpastrana Please rereview

HPCCSmoketest commented:
Automated Smoketest Build: failed
Number of Error(s): 4
/mnt/disk1/home/vamosax/smoketest/smoketest-8381/HPCC-Platform/system/security/plugins/htpasswdSecurity/htpasswdSecurity.cpp:135:88: error: ‘IEspSecureContext’ has not been declared
make[2]: *** [system/security/plugins/htpasswdSecurity/CMakeFiles/htpasswdSecurity.dir/htpasswdSecurity.cpp.o] Error 1
make[1]: *** [system/security/plugins/htpasswdSecurity/CMakeFiles/htpasswdSecurity.dir/all] Error 2
make: *** [all] Error 2

rpastrana commented:
@RussWhitehead most recent changes seem reasonable.

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

RussWhitehead commented:
@richardkchapman Please merge and close

RussWhitehead committed:
HPCC-15047 Refactor HTPass into an embedded SecManager plugin

Refactor HTPasswrd security manager to be a true plugin. Add files for
configmanager to gather common plugin artifacts and custom HTPasswd
specific, rework ISecManager factory to conform to new interface

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

RussWhitehead commented:
@rpastrana Please review last change for "authorize" signature

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on system/security/plugins/htpasswdSecurity/htpasswdSecurity.cpp:
```diff
@@ -122,7 +122,7 @@ class CHtpasswdSecurityManager : public CBaseSecurityManager
     }
 
-    bool authorize(ISecUser & user, ISecResourceList * resources)
+    bool authorize(ISecUser & user, ISecResourceList * resources, IEspSecureContext* secureContext)
```
we should make use of the c11 override specifier

rpastrana commented:
@RussWhitehead one comment, otherwise it looks fine.

RussWhitehead committed:
Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

RussWhitehead commented:
@richardkchapman Please merge

HPCCSmoketest commented:
Automated Smoketest Build: success
Install hpccsystems-platform-community_6.0.0-beta1.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on esp/bindings/http/platform/httpbinding.cpp:
```diff
@@ -186,9 +186,11 @@ EspHttpBinding::EspHttpBinding(IPropertyTree* tree, const char *bindname, const
-                    Owned<IPropertyTree> secMgrs;
-                    VStringBuffer sm("SecurityManagers/SecurityManager[@name='%s']/HtpasswdSecurityManager", m_authmethod.str());
-                    secMgrCfg.setown(proc_cfg->getPropTree(sm.str()));
+                    VStringBuffer sb("SecurityManagers/SecurityManager[@name='%s']", m_authmethod.str());
```
@timothyklemm beware this change will affect you...
